sr_edc_ethercat_drivers/src/sr10.cpp

```diff
@@ -0,0 +1,573 @@
+
+  // Initialise the publisher for the extra analog inputs, gyroscope and accelerometer on the palm
+  extra_analog_inputs_publisher.reset(
+          new realtime_tools::RealtimePublisher<std_msgs::Float64MultiArray>(nodehandle_, "palm_extras", 10));
+
+  // Debug real time publisher: publishes the raw ethercat data
+  debug_publisher = boost::shared_ptr<realtime_tools::RealtimePublisher<sr_robot_msgs::EthercatDebug> >(
+          new realtime_tools::RealtimePublisher<sr_robot_msgs::EthercatDebug>(nodehandle_, "debug_etherCAT_data", 4));
+
+
+  std::string imu_name = device_joint_prefix_ + "imu";
+  imu_state_ = hw_->getImuState(imu_name);
+
+  // Broadcast /set_gyr_scale and /set_acc_scale services
+  imu_gyr_scale_server_ =  nodehandle_.advertiseService
+    <sr_robot_msgs::SetImuScale::Request, sr_robot_msgs::SetImuScale::Response>
+    ("/" + imu_name + "/set_gyr_scale", boost::bind(&SR10::imu_scale_callback_, this, _1, _2, "gyr"));
+  imu_acc_scale_server_ =  nodehandle_.advertiseService
+    <sr_robot_msgs::SetImuScale::Request, sr_robot_msgs::SetImuScale::Response>
+    ("/" + imu_name + "/set_acc_scale", boost::bind(&SR10::imu_scale_callback_, this, _1, _2, "acc"));
+
+  // Initialize default IMU scaling values
+  ros::param::param<int>("/" + imu_name + "/acc_scale", imu_scale_acc_, 0);
+  ros::param::param<int>("/" + imu_name + "/gyr_scale", imu_scale_gyr_, 0);
+
+  // and set the flag to true so that the IMU data is scaled
+  imu_scale_change_ = true;
+
+  return retval;
+}
+
+/**
+  *  This function provides diagnostics data that can be displayed by
+  *  the runtime_monitor node. We use the mutliDiagnostics as it publishes
+  *  the diagnostics for each motors.
+  *
+  * @param diagnostic_vector the vector of diagnostic messages
+  * @param buffer is a pointer to some string. Not used in SR10.
+  *        Can we remove this? It would involve also removing
+  *        it from ethercat_hardware (EthercatHardwareDiagnosticsPublisher::publishDiagnostics)
+  */
+void SR10::multiDiagnostics(vector<diagnostic_msgs::DiagnosticStatus> &diagnostic_vector, unsigned char *buffer)
+{
+  diagnostic_updater::DiagnosticStatusWrapper &diagnostic_status(diagnostic_status_);
+
+  stringstream name;
+  string prefix = device_id_.empty() ? device_id_ : (device_id_ + " ");
+  diagnostic_status.name = prefix + "EtherCAT Dual CAN Palm";
+  diagnostic_status.summary(diagnostic_status.OK, "OK");
+  stringstream hardware_id;
+  hardware_id << sh_->get_product_code() << "-" << sh_->get_serial();
+  diagnostic_status.hardware_id = hardware_id.str();
+
+  diagnostic_status.clear();
+  diagnostic_status.addf("Position", "%02d", sh_->get_ring_position());
+  diagnostic_status.addf("Product Code", "%d", sh_->get_product_code());
+  diagnostic_status.addf("Serial Number", "%d", sh_->get_serial());
+  diagnostic_status.addf("Revision", "%d", sh_->get_revision());
+  diagnostic_status.addf("Counter", "%d", ++counter_);
+
```

Comment:
Let's add  here a `Bad Frame Counter`

```
diagnostic_status.addf("Bad Frame Counter", "%d", bad_frame_counter_);
```

Reply:
Removed for now, as this changes will be part of a separate PR: https://github.com/shadow-robot/sr-ros-interface-ethercat/pull/265/commits/8dffda4fd7ea2bdbc0e483260b80d1107dde9176